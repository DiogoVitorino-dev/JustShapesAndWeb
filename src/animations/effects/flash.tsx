import React, { useEffect } from "react";
import { ColorValue, StyleSheet } from "react-native";
import Animated, {
  useAnimatedStyle,
  useSharedValue,
  withDelay,
  withRepeat,
  withSequence,
  withTiming,
  cancelAnimation as cancel,
  runOnUI,
  runOnJS,
} from "react-native-reanimated";

import { AnimationEffectProps, RunnableAnimation } from "../animations.type";

import Colors from "@/constants/Colors";

export interface FlashConfig {
  duration?: number;
  intensity?: number;
  color?: ColorValue;
  numbersOfReps?: number;
  delayOfReps?: number;
}

export interface FlashProps
  extends RunnableAnimation,
    AnimationEffectProps,
    FlashConfig {}

type FlashEffect = (flash: FlashProps) => React.JSX.Element;

export const Flash: FlashEffect = ({
  color = Colors.effects.flash,
  delayOfReps = 0,
  intensity = 40,
  duration = 200,
  numbersOfReps = 1,
  start,
  onFinish,
  children,
  style,
  view,
}) => {
  intensity = intensity / 100;
  const opacity = useSharedValue(0);

  const animatedStyle = useAnimatedStyle(() => ({
    backgroundColor: color,
    opacity: opacity.value,
  }));

  const handleOnFinished = (finished?: boolean) => {
    "worklet";
    if (finished && onFinish) {
      runOnJS(onFinish)();
    }
  };

  const repeated = () => {
    "worklet";
    if (numbersOfReps !== 1 && numbersOfReps !== 0) {
      opacity.value = withRepeat(
        withDelay(
          delayOfReps,
          withSequence(
            withTiming(intensity, { duration: duration / 2 }),
            withTiming(0, { duration: duration / 2 }),
          ),
        ),
        numbersOfReps - 1,
        false,
        handleOnFinished,
      );
    } else {
      handleOnFinished(true);
    }
  };

  const startAnimation = () => {
    "worklet";
    opacity.value = withSequence(
      withTiming(intensity, { duration: duration / 2 }),
      withTiming(0, { duration: duration / 2 }, (fin) =>
        fin ? repeated() : undefined,
      ),
    );
  };

  const cancelAnimation = () => {
    "worklet";
    cancel(opacity);
    opacity.value = 0;
  };

  useEffect(() => {
    if (start) {
      runOnUI(startAnimation)();
    } else {
      runOnUI(cancelAnimation)();
    }
  }, [start]);

  return (
    <Animated.View style={[styles.container, style]} {...view}>
      {children}
      <Animated.View
        testID="flashAnimation"
        style={[styles.flash, animatedStyle]}
      />
    </Animated.View>
  );
};

const styles = StyleSheet.create({
  container: {
    width: "100%",
    flex: 1,
  },
  flash: {
    width: "100%",
    height: "100%",
    position: "absolute",
    zIndex: 10,
  },
});
