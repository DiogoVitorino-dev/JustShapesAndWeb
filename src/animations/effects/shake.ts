import {
  Easing,
  WithSpringConfig,
  WithTimingConfig,
  useAnimatedStyle,
  useSharedValue,
  withSequence,
  withSpring,
  withTiming,
} from "react-native-reanimated";

import { RunnableAnimation, StylizedAnimation } from "../animations.type";

import { AnimatedStyleApp } from "@/constants/commonTypes";
import { useTimeoutValue } from "@/hooks";
import { MathUtils } from "@/utils/mathUtils";

export type ShakeImpact = "start" | "end" | "all";

export interface ShakeImpactDirection {
  /**
   * Sets the impact side to left (Start), right (End), or both sides (All).
   */
  horizontal?: ShakeImpact;
  /**
   * Sets the impact side to up (Start), down (End), or both sides (All).
   */
  vertical?: ShakeImpact;
}

export interface ShakeImpactConfig extends ShakeImpactDirection {
  /**
   * Multiplier that defines the frequency of impacts
   */
  frequency?: number;
}

export interface ShakeAnimation extends StylizedAnimation, RunnableAnimation {}

type ShakeAnimationProps = Required<Parameters<typeof useShakeAnimation>>;

const InitialShakeImpactConfig: Required<ShakeImpactConfig> = {
  frequency: 1,
  horizontal: "all",
  vertical: "all",
};

const [InitialDuration, InitialAmount, InitialImpact]: ShakeAnimationProps = [
  200,
  20,
  InitialShakeImpactConfig,
];
/**
 *
 * @param duration Duration of the entire animation.
 * @param amount The maximum value that will determine how much the impact will displace, in pixels.
 * @param impact Setting for the impacts.
 * @returns An `animatedStyle` and `run` function to start the animation. Note: `run` function also returns the `animatedStyle`.
 */

export function useShakeAnimation(
  duration: number = InitialDuration,
  amount: number = InitialAmount,
  impact: ShakeImpactConfig = InitialImpact,
): ShakeAnimation {
  const shakeY = useSharedValue(0);
  const shakeX = useSharedValue(0);
  const timer = useTimeoutValue(duration);
  impact = { ...InitialImpact, ...impact };

  const animatedStyle: AnimatedStyleApp = useAnimatedStyle(() => ({
    transform: [{ translateX: shakeX.value }, { translateY: shakeY.value }],
  }));

  const run = () => {
    if (timer.value) {
      const springConfig: WithSpringConfig = {
        duration: duration / 2,
        dampingRatio: 2,
      };

      const timingConfig: WithTimingConfig = {
        duration: duration / 2,
        easing: Easing.in(Easing.bounce),
      };

      const repeating = setInterval(() => {
        shakeY.value = withSequence(
          withSpring(selectDirection(impact.vertical), springConfig),
          withTiming(0, timingConfig),
        );

        shakeX.value = withSequence(
          withSpring(selectDirection(impact.horizontal), springConfig),
          withTiming(0, timingConfig),
        );
      }, duration / impact!.frequency!);

      timer.run();
      setTimeout(() => {
        clearInterval(repeating);
      }, duration);
    }

    return { animatedStyle };
  };

  const selectDirection = (direction: ShakeImpact = "all") => {
    const { random } = MathUtils;

    switch (direction) {
      case "start":
        return random(0, amount * -1);

      case "end":
        return random(0, amount);

      default:
        return random(amount * -1, amount);
    }
  };

  return { animatedStyle, run };
}
