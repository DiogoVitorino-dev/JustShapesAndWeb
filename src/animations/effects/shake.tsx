import React, { useEffect } from "react";
import Animated, {
  Easing,
  WithSpringConfig,
  WithTimingConfig,
  cancelAnimation,
  runOnJS,
  useAnimatedStyle,
  useSharedValue,
  withDelay,
  withSequence,
  withSpring,
  withTiming,
} from "react-native-reanimated";

import { AnimatedEffectProps, RunnableAnimation } from "../animations.type";

import { AnimatedStyleApp } from "@/constants/commonTypes";
import { MathUtils } from "@/utils/mathUtils";

export type ShakeImpact = "none" | "start" | "end" | "all";

export interface ShakeImpactDirection {
  /**
   * Sets the impact side to left (Start), right (End), both sides (All) or none.
   */
  horizontal?: ShakeImpact;
  /**
   * Sets the impact side to up (Start), down (End), both sides (All) or none.
   */
  vertical?: ShakeImpact;
}

export interface ShakeImpactConfig extends ShakeImpactDirection {
  /**
   * Multiplier that defines the frequency of impacts
   */
  frequency?: number;
}

export interface ShakeConfig {
  /**
   * Duration of the entire animation.
   */
  duration?: number;
  /**
   * The maximum value that will determine how much the impact will displace, in pixels.
   */
  amount?: number;
  /**
   * Setting for the impacts.
   */
  impact?: ShakeImpactConfig;
  /**
   * Sets a delay for the effect to start
   */
  delay?: number;
}

const InitialShakeImpactConfig: Required<ShakeImpactConfig> = {
  frequency: 1,
  horizontal: "all",
  vertical: "all",
};

export interface ShakeProps
  extends ShakeConfig,
    RunnableAnimation,
    AnimatedEffectProps {}

type ShakeEffect = (props: ShakeProps) => React.JSX.Element;

export const Shake: ShakeEffect = ({
  amount = 20,
  delay = 0,
  duration = 200,
  impact = InitialShakeImpactConfig,
  start = false,
  view,
  children,
  onFinish,
  style,
}) => {
  const shakeY = useSharedValue(0);
  const shakeX = useSharedValue(0);

  impact = { ...InitialShakeImpactConfig, ...impact };

  const shakeDuration = 80 / impact!.frequency!;

  const timingConfig: WithTimingConfig = {
    duration: shakeDuration,
    easing: Easing.out(Easing.ease),
  };

  const springConfig: WithSpringConfig = {
    stiffness: 200,
  };

  const animatedStyle: AnimatedStyleApp = useAnimatedStyle(() => ({
    transform: [{ translateX: shakeX.value }, { translateY: shakeY.value }],
  }));

  const run = () => {
    let frequency = Number((duration / shakeDuration).toFixed(0)) - 1;

    const sequenceX: number[] = [];
    const sequenceY: number[] = [];

    while (frequency > 0) {
      frequency--;

      sequenceY.push(
        withTiming(selectDirection(impact.vertical), timingConfig),
      );
      sequenceX.push(
        withTiming(selectDirection(impact.horizontal), timingConfig),
      );

      if (frequency === 0) {
        sequenceX.push(withSpring(0, springConfig));
        sequenceY.push(
          withSpring(0, springConfig, (fin) =>
            fin && onFinish ? runOnJS(onFinish)() : undefined,
          ),
        );
      }
    }

    shakeY.value = withDelay(delay, withSequence(...sequenceY));
    shakeX.value = withDelay(delay, withSequence(...sequenceX));
  };

  const cancel = () => {
    cancelAnimation(shakeX);
    cancelAnimation(shakeY);
    shakeX.value = withSpring(0, springConfig);
    shakeY.value = withSpring(0, springConfig, (fin) =>
      fin && onFinish ? runOnJS(onFinish)() : undefined,
    );
  };
  const selectDirection = (direction: ShakeImpact = "all") => {
    const { random } = MathUtils;

    switch (direction) {
      case "start":
        return random(0, amount * -1);

      case "end":
        return random(0, amount);

      case "all":
        return random(amount * -1, amount);

      default:
        return 0;
    }
  };

  useEffect(() => {
    if (start) run();

    return () => {
      if (start) cancel();
    };
  }, [start]);

  return (
    <Animated.View style={[animatedStyle, style]} {...view}>
      {children}
    </Animated.View>
  );
};
