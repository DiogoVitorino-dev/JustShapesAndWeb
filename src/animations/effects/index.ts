import { Flash } from "./flash";
import { Shake } from "./shake";

export const AnimationEffects = {
  Shake,
  Flash,
};
