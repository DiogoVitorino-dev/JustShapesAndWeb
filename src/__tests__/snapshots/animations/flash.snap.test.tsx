import { render } from "@testing-library/react-native";

import { Flash } from "@/animations/effects/flash";

describe("Grenade - Snapshot animation test", () => {
  it("should render correctly", () => {
    const tree = render(<Flash start={false} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
