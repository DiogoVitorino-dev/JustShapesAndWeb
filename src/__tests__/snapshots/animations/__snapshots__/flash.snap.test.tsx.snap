// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grenade - Snapshot animation test should render correctly 1`] = `
<View
  animatedStyle={
    {
      "value": {},
    }
  }
  collapsable={false}
  style={
    {
      "flex": 1,
      "width": "100%",
    }
  }
>
  <View
    animatedStyle={
      {
        "value": {
          "backgroundColor": "rgb(245, 245, 245)",
          "opacity": 0,
        },
      }
    }
    collapsable={false}
    style={
      {
        "backgroundColor": "rgb(245, 245, 245)",
        "height": "100%",
        "opacity": 0,
        "position": "absolute",
        "width": "100%",
        "zIndex": 10,
      }
    }
    testID="flashAnimation"
  />
</View>
`;
